### PR TITLE
Full CI support for public builds + switch to use cibuildwheel

### LINKED ISSUE #227: Validate wheel build in CI/CD
We need to build wheels in the public CI because
- it generates a log that can be audited by non-team members when an issue happens (ex: #226)
- the artifacts can be inspected and reused (ex: moving to internal artifactory for QA/release processes)
- wheel contents can be tested as part of CI

### PR DESCRIPTION
**Pull request by leofang**
Close #227. Part of #81.

The earlier build pipeline was deficient because
1. it only covered a single point in our large support matrix (ex: no aarch64/win-64 coverage)
2. it used a miniforge container to build wheels, which does not guarantee compatibility with the manylinux/glibc requirement
3. since a container is used, it is impossible to reuse many widely used/validated, OSS GHA workflows, and everything has to be homegrown scripts, which is an unnecessary burden to both developer and devops teams
   - one addition downside of this is it's really hard to understand the overall CI design, since codes are scattered into many files that are hard to follow
4. the old CI had too many unnecessary pieces not suitable for this project (ex: CMake, Legate, ...)

This PR switches to use public GHA workflows. Most steps are meant to be easily reusable (after some tweaking/refactoring) by other Python/CUDA projects, either first-party or third-party. For example, `cibuildwheels` has been used to offer guarantee to manylinux. It also lays the ground work for later bring-up of the CI test stage. (This PR only covers the build stage.) 

As of commit https://github.com/NVIDIA/cuda-python/commit/0d554f0945b1bd1b7797170ece2a13765ab6dbe8, we can build all needed wheels using GHA now! 🎉
- `cuda.bindings`
- `cuda.core`
- Python: 3.9-3.12
- OS: linux-64, linux-aarch64, win-64
- CUDA 12.6 (we can evaluate backporting to the cuda-11 branch)
 
The artifacts are available here for inspection: https://github.com/NVIDIA/cuda-python/actions/runs/12192916020

### CODE REVIEW

**Comment by ksimpson-work** [comment]
I'm really liking the look of this so far :)

**Comment by leofang** [comment]
OK  we can build on all platforms now! 🎉

**Comment by leofang** [comment]
OK, `cuda.core` is built successfully for all supported Python x OS versions, the artifacts can be retrieved/inspected here:
https://github.com/NVIDIA/cuda-python/actions/runs/12189979543
I am going to work on building `cuda.bindings` (for CUDA 12) next.

**Comment by leofang** [comment]
As of commit 0d554f0, we can build all needed wheels (`cuda.{bindings,core}` x Python 3.9-3.12 x {linux-64, linux-aarch64, win-64} x CUDA 12.x) using GHA now! 🎉 Artifacts are available here:
https://github.com/NVIDIA/cuda-python/actions/runs/12192916020

**Comment by leofang** [comment]
(updated the PR description)

File: .github/actions/build/action.yml [hunk 1]
```diff
@@ -13,73 +13,79 @@ inputs:
...
+      env:
+        CIBW_BUILD: ${{ env.CIBW_BUILD }}
+        CIBW_ARCHS_LINUX: "native"
+        CIBW_BUILD_VERBOSITY: 1
+        CIBW_ENVIRONMENT_LINUX: >
+          CUDA_PATH="$(realpath ./cuda_toolkit)"
+          PARALLEL_LEVEL=${{ env.PARALLEL_LEVEL }}
+        CIBW_ENVIRONMENT_WINDOWS: >
+          CUDA_HOME="$(cygpath -w $(realpath ./cuda_toolkit))"
+        #  PARALLEL_LEVEL=${{ env.PARALLEL_LEVEL }}
```
Comment (leofang) [inline .github/actions/build/action.yml:66]: Note: This was commented out because of #271.

File: .github/actions/setup/action.yml [hunk 2]
```diff
@@ -22,51 +22,179 @@ inputs:
...
   steps:
-    - name: Set REPO_DIR and Dump environment
+    # WAR: setup-python is not relocatable...
+    # see https://github.com/actions/setup-python/issues/871
+    - name: Set up Python ${{ inputs.python-version }}
+      if: ${{ startsWith(inputs.host-platform, 'linux') }}
+      id: setup-python
+      uses: actions/setup-python@v5
+      with:
+        python-version: "3.12"
```
Comment (leofang) [inline .github/actions/setup/action.yml:39]: Technically we don't need this step. However, because `actions/setup-python` is not relocatable (its RPATH is hard-wired, see https://github.com/actions/setup-python/issues/871), on a self-hosted runner cibuildwheels cannot properly load libpython3.12. This is a WAR by forcing the same Python version installed on the host (and mounted into the container).

File: .github/actions/setup/action.yml [hunk 3]
```diff
@@ -22,51 +22,179 @@ inputs:
...
+    # - name: Set up new PowerShell (pwsh)
+    #   if: ${{ startsWith(inputs.host-platform, 'win') }}
+    #   shell: powershell
+    #   env:
+    #     # doesn't seem there's an easy way to avoid hard-coding it?
+    #     PS_MSI_URL: https://github.com/PowerShell/PowerShell/releases/download/v7.4.6/PowerShell-7.4.6-win-x64.msi
+    #   run: |
+    #     Invoke-WebRequest -Uri "$env:PS_MSI_URL" -OutFile "ps_installer.msi"
+    #     Start-Process msiexec.exe -Wait -Verbose -ArgumentList '/i "ps_installer.msi" /qn REGISTER_MANIFEST=1'
+    #     echo "$env:ProgramFiles\\PowerShell\\7" >> $env:GITHUB_PATH
```
Comment (leofang) [inline .github/actions/setup/action.yml]: We can clean this up now, as it's now tracked in https://github.com/nv-gha-runners/vm-images/issues/102. (Our images do not contain pre-installed `pwsh`.)

File: .github/actions/setup/action.yml [hunk 4]
```diff
@@ -22,51 +22,179 @@ inputs:
...
+    #     # # For some reason below just doesn't work in the CI...
+    #     # Start-Process -NoNewWindow -RedirectStandardError cl.out -FilePath "$clPath"
+    #     # $CL_VERSION_STRING = & type cl.out
+    #     # if (($CL_VERSION_STRING -join " ") -match "Version (\\d+\\.\\d+)\\.\\d+") {
+    #     #     $CL_VERSION = $matches[1]
+    #     #     echo "Detected cl.exe version: $CL_VERSION"
+    #     # }
+
+    #     # Delete temporary items created by installers and such
+    #     Remove-Item -Recurse "${env:TEMP}\\*"
```
Comment (leofang) [inline .github/actions/setup/action.yml]: We can clean this up now, as it's now tracked in https://github.com/nv-gha-runners/vm-images/issues/103. (Our images do not contain pre-installed MSVC.)

File: .github/actions/setup/action.yml [hunk 5]
```diff
@@ -22,51 +22,179 @@ inputs:
...
+        elif [[ "${{ inputs.host-platform }}" == "linux-aarch64" ]]; then
           TARGET_PLATFORM='linux-aarch64'
+        elif [[ "${{ inputs.host-platform }}" == "win-x64" ]]; then
+          TARGET_PLATFORM='win-64'
         fi
 
-        BUILD_MODE="${{ inputs.build-mode }}"
-        BUILD_MODE_STR=""
-        [ -n "${BUILD_MODE}" ] && BUILD_MODE_STR="-${BUILD_MODE}"
+        if [[ "${{ inputs.host-platform }}" == linux* ]]; then
```
Comment (keenan-simpson) [inline .github/actions/setup/action.yml:115]: nit: we could format the python version instead of doing the if else fallthrough:

```yml
        CIBW_BUILD='cp$(echo '${{ inputs.python-version }}' | tr -d '.')-manylinux*'

        CIBW_BUILD='cp$(echo '${{ inputs.python-version }}' | tr -d '.')-win_amd64*'
```

Comment (ksimpson-work) [inline .github/actions/setup/action.yml:157]:
```suggestion
        CIBW_BUILD='cp{$(echo '${{ inputs.python-version }}' | tr -d '.')}-manylinux*'
```

File: .github/workflows/gh-build-and-test.yml [hunk 6]
```diff
@@ -1,16 +1,16 @@
 on:
   workflow_call:
     inputs:
-      host-platform:
```
Comment (keenan-simpson) [inline .github/workflows/gh-build-and-test.yml:4]: This could be saved for the next PR, but having this called build-and-test has been misleading imo. I think it could be better to call gh-build.yml and gh-test.yml from ci-gh.yml

File: .github/actions/setup/action.yml [hunk 7]
```diff
@@ -22,51 +22,179 @@ inputs:
...
+    # WAR: setup-python is not relocatable...
+    # see https://github.com/actions/setup-python/issues/871
+    - name: Set up Python ${{ inputs.python-version }}
+      if: ${{ startsWith(inputs.host-platform, 'linux') }}
+      id: setup-python
+      uses: actions/setup-python@v5
+      with:
+        python-version: "3.12"
+
+    # - name: Set up new PowerShell (pwsh)
```
Comment (keenan-simpson) [inline .github/actions/setup/action.yml]: My gut tells me we should have the windows stuff uncommented before approval, or take all windows related changes out of this review. Having the install-vs script included with this commented out for example. Per your recent comments it seems this is being address by the nv-gha-runnners issues. Do we want to consider those dependencies?

Comment (keenan-simpson) [reply .github/actions/setup/action.yml]: nvm per plan to switch to nv pre installed runners. Will wait to approve until that stuff is removed?

File: .github/workflows/gh-build-and-test.yml (hunk 6 above)
Comment (leofang) [reply .github/workflows/gh-build-and-test.yml:4]: Discussed on the call, yes we'll have to refactor this to accommodate that we have different build/test matrices to cover.

File: .github/actions/setup/action.yml (hunk 7 above)
Comment (leofang) [reply .github/actions/setup/action.yml]: - Using GH-hosted runners: pwsh and msvc are both included in the Windows images.
- Using self-hosted (nv-gha) runners: Yes we need them to support both, this is currently a blocker tracked in both vm-images issues.

File: .github/actions/setup/action.yml (hunk 3 above)
Comment (leofang) [reply .github/actions/setup/action.yml]: removed

File: .github/actions/setup/action.yml (hunk 4 above)
Comment (leofang) [reply .github/actions/setup/action.yml]: removed in [19445d5](https://github.com/NVIDIA/cuda-python/pull/267/commits/19445d58143f0dc58193a39080c5808a4e52d220)

**Comment by ksimpson-work** [comment]
Last comment I think: Perhaps we should squash these commits. Not hellbent on this

File: .github/actions/setup/action.yml [hunk 8]
```diff
@@ -22,51 +22,136 @@ inputs:
...
         if [[ ("${BUILD_MODE}" == "") || ("${BUILD_MODE}" == "release") ]]; then
           # We upload release versions in the default folder.
           PKG_DIR="${TARGET_PLATFORM}"
         else
           PKG_DIR="${BUILD_MODE}/${TARGET_PLATFORM}"
         fi
 
-        PYTHON_VERSION_FORMATTED=$(echo '${{ inputs.python-version }}' | tr -d '.')
-
-        echo "BINDINGS_ARTIFACT_NAME=${{ inputs.host-platform }}-${{ inputs.build-type }}-cuda_bindings-python${PYTHON_VERSION_FORMATTED}-${{ inputs.target-device }}${BUILD_MODE_STR}${WITH_TESTS_STR}-${{ github.sha }}" >> $GITHUB_ENV
```
Comment (keenan-simpson) [inline .github/actions/setup/action.yml:62]: or could do it like this with the PYTHON_VERSION_FORMATTED var

File: .github/actions/setup/action.yml (hunk 5 above)
Comment (leofang) [reply .github/actions/setup/action.yml:115]: Done in 07b8d1d.

File: .github/actions/setup/action.yml (hunk 8 above)
Comment (leofang) [reply .github/actions/setup/action.yml:62]: yup, done in commit 07b8d1d

**Comment by leofang** [comment]
> Perhaps we should squash these commits. 

Yes when we merge there's a drop-down menu to select merge or squash.

**Comment by ksimpson-work** [comment]
ahh, that option is grayed out for me on this project

**Review by ksimpson-work — APPROVED** [review]

**Comment by leofang** [comment]
@ksimpson-work is it still grey after you approved it?

**Comment by ksimpson-work** [comment]
No, I noticed it when I went to merge my own PRs. I'll take a screenshot next time, unless I was hallucinating

**Comment by leofang** [comment]
Let's merge! Thanks for quick reviews @ksimpson-work!
